SOSS/dms/soss_utils.py

```diff
@@ -0,0 +1,164 @@
+    dimy, dimx = np.shape(image)
+
+    # If no header was passed we have to check all possible sizes.
+    if header is None:
+
+        # Initialize padding to zero in this case because it is not a reference file.
+        padding = 0
+
+        # Assume the stack is a valid SOSS subarray.
+        # FULL: 2048x2048 or 2040x2040 (working pixels) or multiple if oversampled.
+        # SUBSTRIP96: 2048x96 or 2040x96 (working pixels) or multiple if oversampled.
+        # SUBSTRIP256: 2048x256 or 2040x252 (working pixels) or multiple if oversampled.
+
+        # Check if the size of the x-axis is valid.
+        if (dimx % 2048) == 0:
+            xnative = 2048
+            xos = int(dimx // 2048)
+
+        elif (dimx % 2040) == 0:
+            xnative = 2040
+            xos = int(dimx // 2040)
+
+        else:
+            raise ValueError('Stack X dimension has unrecognized size of {:}. Accepts 2048, 2040 or multiple of.'.format(dimx))
+
+        # Check if the y-axis is consistent with the x-axis.
+        if np.int(dimy/xos) in [96, 256, 252, 2040, 2048]:
+            yos = np.copy(xos)
+            ynative = np.int(dimy/yos)
+
+        else:
+            raise ValueError('Stack Y dimension ({:}) is inconsistent with X dimension ({:}) for acceptable SOSS arrays'.format(dimy, dimx))
+
+        # Create a boolean mask indicating which pixels are not reference pixels.
+        refpix_mask = np.ones_like(image, dtype='bool')
+        if xnative == 2048:
+            # Mask out the left and right columns of reference pixels.
+            refpix_mask[:, :xos * 4] = False
+            refpix_mask[:, -xos * 4:] = False
+
+        if ynative == 2048:
+            # Mask out the top and bottom rows of reference pixels.
+            refpix_mask[:yos * 4, :] = False
+            refpix_mask[-yos * 4:, :] = False
+
+        if ynative == 256:
+            # Mask the top rows of reference pixels.
+            refpix_mask[-yos * 4:, :] = False
+
+    else:
+        # Read the oversampling and padding from the header.
+        padding = int(header['PADDING'])
+        xos, yos = int(header['OVERSAMP']), int(header['OVERSAMP'])
+
+        # Check that the stack respects its intended format.
+        if (dimx/xos - 2*padding) not in [2048]:
+            raise ValueError('The header passed is inconsistent with the X dimension of the stack.')
+        else:
+            xnative = 2048
+
+        if (dimy/yos - 2*padding) not in [96, 256, 2048]:
+            raise ValueError('The header passed is inconsistent with the Y dimension of the stack.')
+        else:
+            ynative = np.int(dimy/yos - 2*padding)
+
+        # The trace file contains no reference pixels so all pixels are good.
+        refpix_mask = np.ones_like(image, dtype='bool')
+
+    # If verbose print the output.
+    if verbose:
+        print('Data dimensions:')
+        print('dimx={:}, dimy={:}, xos={:}, yos={:}, xnative={:}, ynative={:}'.format(dimx, dimy, xos, yos, xnative, ynative))
```

Comment:
again long lines are hard to read, and easy to split up
```python
pargs = [dimx, dimy, xos, yos, xnative, ynative]
pmsg = 'dimx={:}, dimy={:}, xos={:}, yos={:}, xnative={:}, ynative={:}'
print(pmsg.format(*pargs)
```

Reply:
Fixed.